cucumber-core/src/main/java/io/cucumber/core/runner/CachingGlue.java

```diff
@@ -229,17 +243,40 @@ List<DocStringTypeDefinition> getDocStringTypeDefinitions() {
         return docStringTypeDefinitions;
     }
 
-    void prepareGlue(StepTypeRegistry stepTypeRegistry) throws DuplicateStepDefinitionException {
-        StepExpressionFactory stepExpressionFactory = new StepExpressionFactory(stepTypeRegistry, bus);
+    StepTypeRegistry getStepTypeRegistry() {
+        return stepTypeRegistry;
+    }
+
+    void prepareGlue(Locale locale) throws DuplicateStepDefinitionException {
+        boolean firstTime = stepTypeRegistry == null;
+        boolean languageChanged = !locale.equals(this.locale);
+        boolean mustRebuildCache = false;
+        if (firstTime || languageChanged || dirtyCache || hasScenarioScopedGlue) {
+            // conditions changed => invalidate the glue cache
+            // Note: we have a prudent approach of avoiding caching if
+            // scenario-scoped glue exist (e.g. cucumber-java8).
+            this.locale = locale;
+            stepTypeRegistry = new StepTypeRegistry(locale);
+            stepExpressionFactory = new StepExpressionFactory(stepTypeRegistry, bus);
+            stepDefinitionsByPattern.clear();
+            stepPatternByStepText.clear();
+            mustRebuildCache = true;
+            // since we must rebuild the cache, it will not be dirty the next
+            // time
+            dirtyCache = false;
+        }
 
         // TODO: separate prepared and unprepared glue into different classes
-        parameterTypeDefinitions.forEach(ptd -> {
-            ParameterType<?> parameterType = ptd.parameterType();
-            stepTypeRegistry.defineParameterType(parameterType);
-            emitParameterTypeDefined(ptd);
-        });
-        dataTableTypeDefinitions.forEach(dtd -> stepTypeRegistry.defineDataTableType(dtd.dataTableType()));
-        docStringTypeDefinitions.forEach(dtd -> stepTypeRegistry.defineDocStringType(dtd.docStringType()));
+        if (mustRebuildCache) {
```

Comment:
I think all code in the `prepareGlue` method can be kept within this if statement. Do you agree?

Reply:
You mean, including the before/after hook/step hook event emitting or only the default transformers ? It would be ok to include the default transformers in the `if`. But the before/after hook and step hook event emitting behavior will change: currently, we emit the events for each call to `prepareGlue` (that is every time we run a pickle, if I'm right). If we enclose the emit throwing into the `if (mustRebuildCache)` guard, then we will only emit events only when the cache is refreshed. This makes the `compatibility` checks to fail.

Reply:
Okay. I'll look at that.

Reply:
Only emitting messages when the cache is updated is the expected behavior. Otherwise we're emitting duplicate messages for hooks and step definitions that have already been defined.

Reply:
I think the failure was caused by non-determinism around the ordering of hooks.